### PR TITLE
Adds perennial crop model

### PR DESCRIPTION
**Pull request by evasinha**
- Adds perennial crop model and expands crop types to 36.
- New parameter file was generated for 36 CFT's and for parameters for perennial crops.

[BFB] except for crop tests.

### CODE REVIEW

**Review by bbye — APPROVED** [review]

File: components/elm/src/biogeochem/PhenologyMod.F90 [hunk 1]
```diff
@@ -1910,6 +1930,218 @@ subroutine CropPhenology(num_pcropp, filter_pcropp, &
...
+                 a10tmin(p) > minplanttemp(ivt(p))          .and. &
+                 jday       >= minplantjday(ivt(p),h)       .and. &
+                 jday       <= maxplantjday(ivt(p),h) ) then
+               croplive(p)  = .true.
+               gddmaturity(p) = hybgdd(ivt(p))
+               offset_flag(p) = 1._r8
+               offset_counter(p) = dt
+               harvdate(p) = NOT_Harvested
+               harvday(p) = NOT_Harvested
+               write(iulog,*) 'perennial crops live, t10(p), a10tmin(p)', t10(p), a10tmin(p)
```
[the remark this answers is not in the document]
Comment (evasinha) [reply components/elm/src/biogeochem/PhenologyMod.F90]: I have deleted the line.

File: components/elm/src/biogeochem/PhenologyMod.F90 [hunk 2]
```diff
@@ -1910,6 +1930,218 @@ subroutine CropPhenology(num_pcropp, filter_pcropp, &
...
+                  fert(p) = 0._r8
+               else ! continue same fert application every timestep
+                  fert_counter(p) = fert_counter(p) - dt
+               end if
+
+               ! if end of the onset period, reset phenology flags and indices
+               if (onset_counter(p) <= 0.0_r8 .and. &
+                   (t10(p) /= spval  .and. t10(p) < senestemp(ivt(p)))) then
+
+                  write(iulog,*) 'leaves senescence has occurred'
```
[the remark this answers is not in the document]
Comment (evasinha) [reply components/elm/src/biogeochem/PhenologyMod.F90]: I have deleted the line.

File: components/elm/src/biogeochem/PhenologyMod.F90 [hunk 3]
```diff
@@ -1910,6 +1930,218 @@ subroutine CropPhenology(num_pcropp, filter_pcropp, &
...
+               soilt = t_soisno(c,3)
+               if (onset_gddflag(p) == 1.0_r8 .and. soilt > SHR_CONST_TKFRZ) then
+                  onset_gdd(p) = onset_gdd(p) + (soilt-SHR_CONST_TKFRZ)*fracday
+               end if
+
+               ! if accumulated gdd has exceeded gdd required for leaf onset
+               ! then set onset flag
+               ! for perennial crops gddmin tracks gdd required for leaf onset
+               if (onset_gdd(p) > gddmin(ivt(p))) then
+                  write(iulog,*) 'leaves have emerged, onset_gdd(p)', onset_gdd(p)
```
[the remark this answers is not in the document]
Comment (evasinha) [reply components/elm/src/biogeochem/PhenologyMod.F90]: I have deleted the line.

**Comment by evasinha** [comment]
@peterdschwartz - All tests in E3SM-Land-Developer test suite have passed, except the two smallvilleIA tests that have DIFF status likely due to different parameter file usage.This PR is ready to be merged.

**Comment by peterdschwartz** [comment]
Merged to next

**Comment by peterdschwartz** [comment]
These input files are missing from repos so the tests failed on mappy and compy.   Chrysalis hasn't reported but I ran the tests on there yesterday so the files must have already been there.  

```
surfdata_1x1_smallvilleIA_mp50_simyr2000_c210617.nc 
clm_params_c210219.nc
landuse.timeseries_1x1_smallvilleIA_mp50_hist_simyr1850-1855_c210617.nc
```
I think you'll either have to upload these to the online repo's or perhaps ask the machine point-of-contacts to put directly download the data files.

**Comment by evasinha** [comment]
@peterdschwartz  I have added the files to:
/lcrc/group/acme/public_html/inputdata/lnd/clm2/surfdata_map/
and  /lcrc/group/acme/public_html/inputdata/lnd/clm2/paramdata folders on Anvil.

**Comment by peterdschwartz** [comment]
```
 SMS_D_Ld1.ne30pg2_EC30to60E2r2.WCYCL1850.chrysalis_intel.allactive-wcprod
SMS_D_Ld1.ne30pg2_r05_EC30to60E2r2.A_WCYCL1850S_CMIP6.chrysalis_intel
SMS_D_Ln5.conusx4v1_conusx4v1.FC5AV1C-L.chrysalis_intel
SMS_D_Ln5.ne4_ne4.FC5AV1C-L.chrysalis_intel
SMS_D_Ln5.ne4_ne4.FC5AV1C-L.chrysalis_intel.eam-clubb_sp
SMS_D_Ln5.ne4_ne4.FC5AV1C-L.chrysalis_intel.eam-cosplite_nhtfrq5
```
These tests reported Fails on chrysalis.  I ran them using from merge point on next for this PR and with the branch by itself and also got the fails, so I believe they are due to this PR.  After reverting I re-checked two of the cases:
```
  SMS_D_Ld1.ne30pg2_r05_EC30to60E2r2.A_WCYCL1850S_CMIP6.chrysalis_intel (Overall: PASS) details:
  SMS_D_Ln5.ne4_ne4.FC5AV1C-L.chrysalis_intel (Overall: PASS) details:
```

**Comment by rljacob** [comment]
All those have debugging on (the "_D").  There should be a good error message pointing to the problem.   F-cases without debugging passed.

**Comment by peterdschwartz** [comment]
```
2729: forrtl: severe (408): fort: (2): Subscript #1 of the array CSTR has value 1001 which is greater than the upper bound of 1000
2729:
2729: Image              PC                Routine            Line        Source
2729: libpnetcdf.so.3.0  00007FFFF6B14232  for_emit_diagnost     Unknown  Unknown
2729: e3sm.exe           000000000CB0DB2B  pionfget_mod_mp_c         560  pionfget_mod.F90.in
2729: e3sm.exe           000000000CAFB754  pionfget_mod_mp_g         402  pionfget_mod.F90.in
2729: e3sm.exe           00000000082CC2A5  ncdio_pio_mp_ncd_        1399  ncdio_pio.F90.in
2729: e3sm.exe           0000000005024A99  pftvarcon_mp_pftc         595  pftvarcon.F90
2729: e3sm.exe           0000000004BDF92C  elm_initializemod         248  elm_initializeMod.F90
2729: e3sm.exe           0000000004B317DA  lnd_comp_mct_mp_l         237  lnd_comp_mct.F90
2729: e3sm.exe           000000000048E8A7  component_mod_mp_         257  component_mod.F90
2729: e3sm.exe           0000000000437763  cime_comp_mod_mp_        1425  cime_comp_mod.F90
2729: e3sm.exe           00000000004853A9  MAIN__                    122  cime_driver.F90
2729: e3sm.exe           0000000000424A22  Unknown               Unknown  Unknown
2729: libc-2.28.so       00007FFFF37947B3  __libc_start_main     Unknown  Unknown
2729: e3sm.exe           000000000042492E  Unknown               Unknown  Unknown
```
didn't check them all but this same traceback was in a couple of different cases

**Comment by rljacob** [comment]
@dqwu or @jayeshkrishna , do you know the cause of above errors?

**Comment by jayeshkrishna** [comment]
This error points to an invalid memory access (invalid array subscript) while converting data in a C string (read from a file) to a Fortran string inside Scorpio.

@peterdschwartz / @evasinha : Does SMS_D_Ln5.ne4_ne4.FC5AV1C-L reproduce this issue?

**Comment by evasinha** [comment]
Thanks @jayeshkrishna . I did try this case on anvil and was able to reproduce the error message but have not figure what is causing it. The same case without the debug option runs fine and passes all tests.

**Comment by jayeshkrishna** [comment]
ok, I will also take a look at it and let you know what I find.

**Comment by jayeshkrishna** [comment]
@evasinha : Do you expect pftnames /home/climate1/acme/inputdata/lnd/clm2/paramdata/clm_params_c180301.nc to contain more than 25 names (char pftname(pft, ...) , pft = 25)? It looks like you are expecting 50 names (a guess based on the changes here) in the file.

This error is showing up in the debug version of the test (although the issue is still present in the non-debug version of the test, the code does not crash, it ends up reading invalid values).

**Comment by jayeshkrishna** [comment]
I would recommend that you take a look at the change of mxpft in components/elm/src/main/elm_varpar.F90 , did you expect the value to change across all files ? Due to the change the buffer used in /sandbox/jayesh/E3SM_test2/components/elm/src/main/pftvarcon.F90:595 expects 50 strings, although the input file only contains 25 strings.

(PS: In any case Scorpio should not have crashed in this case, I am testing out a branch with fixes to issue a warning to the user in this case. However I would recommend that you check the code change and ensure that its not an unintentional bug in this PR)

**Comment by evasinha** [comment]
@jayeshkrishna The updated parameter file (`clm_params_c210219.n`) has 50 values for pftnames but the old one (`clm_params_c180301.nc`) only has 25 values. I do see that because of the edits that I have made in `components/elm/src/main/elm_varpar.F90` 50 values for pftnames are expected regardless of whether old or new parameter file is being used. If you are available then I would like to schedule a brief call tomorrow or earlier next week to discuss the issue and how best to modify the code to address it.

**Comment by jayeshkrishna** [comment]
Sure, we can set up a call if needed. But would a conditional allocation of the array work for you ?
I have a temporary fix here that seems to work (I haven't tried the complete test suite though) - https://gist.github.com/jayeshkrishna/ce246427ad25bfd5cf7ba810d328d860 . Please take a look and incorporate it in your code as needed.

**Comment by evasinha** [comment]
@jayeshkrishna I used conditional allocation array as suggested by you and `SMS_D_Ln5.ne4_ne4.FC5AV1C-L` test passed on anvil. Thanks for the great suggestion!

I am guessing all tests should pass now. I will run the E3SM-Land-Developer testsuite again just to ensure no tests are failing and then reply back here.

**Comment by evasinha** [comment]
@jayeshkrishna All tests in E3SM-Land-Developer test suite have passed, except the two smallvilleIA tests that have DIFF status likely due to different parameter file usage. I also tested two of the tests that had failed for you earlier and they passed as well:
```
SMS_D_Ld1.ne30pg2_EC30to60E2r2.WCYCL1850.chrysalis_intel.allactive-wcprod
SMS_D_Ln5.ne4_ne4.FC5AV1C-L.chrysalis_intel
```

**Comment by evasinha** [comment]
Great! Thanks @peterdschwartz

**Comment by evasinha** [comment]
@peterdschwartz I found a small bug in the code and I will make a new commit and update this PR. Sorry, for not finding it earlier.

**Comment by peterdschwartz** [comment]
Merged your commit to next

**Comment by peterdschwartz** [comment]
Merged to master and requested Bless for Smallville test.  Other fails seem to be due to other PRs.

**Pull request merged**